pageserver/src/tenant/timeline.rs

```diff
@@ -275,7 +275,7 @@ impl LogicalSize {
 }
 
 pub struct WalReceiverInfo {
-    pub wal_source_connstr: String,
+    pub wal_source: tokio_postgres::Config,
```

Comment:
To be honest, I don't think this is a good substitute. The downsides of this change are:

* The struct itself is bloated (it has multiple vectors) and contains many things not needed in such a simple use case.
* It does some magic that's uncalled for. For instance, it will set a default port 5432 if you don't provide one. This is fine when you're connecting to a regular postgres, but it's anything but useful when you're connecting to something bespoke, because who in their right mind would use a default port in that case?
* It gives you more flexibility than needed, like the ability to set multiple hosts and ports. **It will no longer be possible to extract a connection string** in a straightforward manner (we used to have a hack for that which [I eliminated yesterday](https://github.com/neondatabase/neon/commit/0df3467146f9368d67a145c59949e7f0ee6c01c9)). Furthermore, it will be possible to abuse this and create broken configurations with e.g. 2 hosts and 3 ports which the library can't handle.
* This is an implementation detail that would be hard to get rid of if we ever decide to replace tokio-postgres with anything else (this is not as hypothetical as you might think).

---

In my opinion, it'd be better to create a newtype like this:

```rust
struct ConnectionUrl(url: url::Url);

impl ConnectionUrl {
  fn as_str(&self) -> &str { self.url.as_str() }
}

// Do the validation here.
impl std::str::FromStr for ConnectionUrl { ... }
```

You could reuse [this](https://github.com/neondatabase/neon/blob/0df3467146f9368d67a145c59949e7f0ee6c01c9/control_plane/src/connection.rs#L4).